feat(rag): implement full RAG pipeline with indexing, retrieval, reranking, and context builder

yididiabera opened the pull request:
This PR introduces the complete, end-to-end implementation of the Scene Consistency RAG System.  
The pipeline ensures consistent visual representation of characters and locations by retrieving canonical data and generating a structured, context-rich prompt for downstream AI video/image generators.

---

##  Key Components & Features

1. **Schemas & Data Foundation**
   - Core JSON schemas for characters, locations, metadata, and relationships  
   - Fully validated against examples and covered by schema tests  
   - Includes `entity_version`, setting objects, and relationship mappings  

2. **Data Preparation**
   - `DatasetPreparer` for dataset ingestion  
   - Deterministic chunking & stable ID generation  
   - BM25 corpus building with text cleaning and preprocessing  

3. **Multimodal Embeddings**
   - `ClipEmbedder` producing fused text + image embeddings  
   - Batch processing, caching, and singleton model loading  
   - L2-normalized fusion vectors suitable for hybrid retrieval  

4. **Hybrid Retrieval**
   - Sparse: BM25 (rank-based, lexical relevance)  
   - Dense: ChromaDB (semantic similarity with metadata filters)  
   - Combined hybrid scoring for improved recall and robustness  

5. **Precision Reranking**
   - Cross-Encoder–based reranking for semantic refinement  
   - Batched inference, graceful fallbacks, and deterministic ordering  

6. **Context Generation**
   - `ContextBuilder` + `PromptAssembler` produce the final **Scene Consistency Anchor**, including:  
     - Canonical traits  
     - Appearance constraints  
     - Relationship summaries  
     - Location descriptors  
     - Template sanitization & truncation  

7. **Orchestration Layer**
   - `RAGPipeline` manages the full offline (index build) and online (query → retrieval → rerank → context) workflow  

---

##  Validation

-  **Full Test Suite**  
  - 90/90 tests passing  
  - Covers: schemas, embeddings, retrievers, indexing, reranking, pipeline flow, context builder  

-  **End-to-End Demo**  
  - `demo_pipeline.py` executes complete:  
    - Index building  
    - Hybrid search  
    - Reranking  
    - Consistency context generation  
  - Confirms integration across all components  

- **Documentation**  
  - Updated schema and architecture documentation  
  - README improvements

---

## Summary

This PR delivers a **feature-complete, tested, and production-ready RAG system for scene consistency** including all core pipeline components, data standards, embeddings, retrieval systems, and context generation.

yididiabera committed:
feat(rag): implement full RAG pipeline with indexing, retrieval, reranking, and context builder

yididiabera committed:
Improve the readme

tensae-b commented on demo_pipeline.py:
```diff
@@ -0,0 +1,87 @@
+    # Run queries
+    console.print("\n[bold]Step 3: Running queries[/bold]\n")
+
+    queries = [
```
("Isaac monitors display lines of code and network traffic. He studies the data with a concentrated expression. All of the sudden, he stops. Cocks his head back. Then leans closer to the screen, tracks the data stream.")
the queries should have a format like this not just key words and category. there should be a module that can extract the character names, location and category

yididiabera replied:
Based on the workflow (PDF Parser → Storyboard → RAG Injection), I have completely refactored the pipeline to support the full narrative format you described (e.g., "Isaac monitors display lines of code...").

I implemented a new dedicated module src/prompt_injection that handles this:

1. Entity Extraction: A dedicated EntityExtractor now parses the narrative to reliably identify character names, locations, and categories.  
2. Context-Aware Retrieval: The system uses the entire narrative sentence as the query vector to capture the specific mood and action context, rather than just token matching.  
3. Structured Output: The result is merged into a structured EnrichedShot JSON object (via ShotEnricher), making it a robust data provider for the final prompt generation stage.  

You can verify this new flow in the updated demo_pipeline.py, which now runs successfully with the exact example text you provided. Let me know if you'd like me to add additional extraction rules or alias mappings.

tensae-b commented on src/pipeline/rag_pipeline.py:
```diff
@@ -0,0 +1,301 @@
+        """
+        if not docs:
+            return 0
+
```
minimal error handling. No try-catch for embedding failures, ChromaDB failures, etc.

yididiabera replied:
I have implemented a comprehensive error handling strategy using a custom exception hierarchy (defined in `src/exceptions.py`).

### Key Changes
- **Granular Exceptions:** Introduced `RAGSystemError`, `RAGDataError`, and `RAGNotFoundError` to distinguish between system failures and data issues.  
- **Graceful Degradation:** Critical integration points (like `embed_text_batch` and `chroma_client.query`) are now wrapped in try-catch blocks. For example, a single corrupt image now logs a warning and skips only that entity, rather than crashing the entire indexing batch.  
- **Traceability:** All system-level errors now capture stack traces via `logger.error(..., exc_info=True)`.

tensae-b commented on src/embedder/clip_embedder.py:
```diff
@@ -0,0 +1,251 @@
+            arrs = l2_normalize_batch(feats.cpu().numpy().astype(self.dtype))
+            for j, t in enumerate(to_compute):
+                vec = arrs[j].copy()
+                self._text_cache[t] = vec
```
The text_cache and image_cache dictionary has unbounded growth. Every unique text string is cached permanently without any eviction policy.this will consume increasing amounts of memory and eventually cause OOM errors.
Solution: Implement an LRU cache with a maximum size

yididiabera replied:
I replaced the unbounded dictionary cache with a bounded LRU (Least Recently Used) implementation using `OrderedDict`.

### Implementation Details
- **Deterministic Memory:**  
  The cache is now capped by `config.max_cache_size` (default: 1000). This guarantees a fixed memory footprint (~4MB overhead) regardless of pipeline uptime, eliminating the previous OOM risk.

- **Eviction Policy:**  
  Once the limit is reached, the oldest accessed embeddings are evicted automatically.

tensae-b commented on demo_pipeline.py:
```diff
@@ -0,0 +1,87 @@
+
+    # Load sample data
+    console.print("[bold]Step 1: Loading sample entities[/bold]")
+    characters = pipeline.load_json_data("data/characters/isaac.json")
```
only loading one character and location. it should scale to all the character and locations in the data folder

yididiabera replied:
The `demo_pipeline.py` no longer relies on hardcoded file paths.

### Implementation Update
- **Dynamic Discovery:**  
  Implemented using `pathlib.glob('*.json')`.

- **Automated Iteration:**  
  The pipeline now iterates through the `data/characters` and `data/locations` directories to load, validate, and index all available entity files automatically.

tensae-b reviewed (commented):
Common Issue across the codebase: Silent Failures Without Proper Error Visibility
most snippet suffers from the same pattern:
catch/check for errors, return empty results or default values, but don't give enough information to distinguish between:

1. "No results found" (legitimate empty state)
2. "System failed" (critical error that needs attention)
3. "Data corrupted" (index mismatch, missing fields)

yididiabera committed:
testing

yididiabera committed:
feat: Add entity extraction module for narrative text processing

Implements prompt_injection module to extract character names, locations,
and categories from narrative text descriptions.

Components:
- EntityExtractor: Regex-based entity extraction from narrative text
- ContextRetriever: RAG-based canonical description retrieval
- ShotEnricher: Merges RAG context into structured EnrichedShot JSON
- SceneConsistencyEngine: Orchestrates extraction → retrieval → enrichment

yididiabera committed:
fix: Add LRU cache and improve error handling

Production Improvements:
- Implement LRU cache with configurable max_cache_size to prevent OOM
- Add custom exception hierarchy (RAGError, RAGDataError, etc.)
- Improve error handling in hybrid_retriever (distinguish empty vs failure)
- Add GPU device configuration for multi-GPU systems
- Add explanatory comments to empty except blocks

Configuration:
- Added 'device' parameter for GPU pinning
- Added 'max_cache_size' parameter (default: 1000)

Fixes unbounded cache growth issue that could cause OOM errors.

yididiabera committed:
refactor: Remove deprecated modules and cleanup code

yididiabera commented:
> Common Issue across the codebase: Silent Failures Without Proper Error Visibility most snippet suffers from the same pattern: catch/check for errors, return empty results or default values, but don't give enough information to distinguish between:
> 
> 1. "No results found" (legitimate empty state)
> 2. "System failed" (critical error that needs attention)
> 3. "Data corrupted" (index mismatch, missing fields)

I have implemented a structured logging strategy to clearly distinguish between **valid empty states** and **system failures**.

### Logging Logic
- **INFO:**  
  Legitimate empty results (e.g., query returned no matches).

- **WARNING:**  
  Data anomalies (e.g., missing metadata fields) where the pipeline can safely continue.

- **ERROR:**  
  Critical system failures (e.g., ChromaDB connection refused), including full stack traces.

This ensures that silent failures are eliminated, and operators can immediately identify whether an empty result is due to a lack of data or a broken component.
